File: compare_files.py
#!/usr/bin/env python
# $Id: compare_files.py 27007 2013-12-02 15:53:34Z amkjpf $
# ==============================================================================
# Name    : compare_files.py
# Desc    : A Python program to compare two csv files in a highly customized
#           fashion.
# Usage   : python compare_files.py -f file1.csv,file2.csv -k <key_list> \
#                                   -p <precision> -x <exclude_list>
# Example : python compare_files.py -f sybase.csv,oracle.csv \
#                                   -k FUNDCODE,FUNDNAME -p 0.01 -x CAPITAL
# By      : prat
# On      : 11/18/2013
# ==============================================================================
import sys
import optparse
keycols = {}    # a dictionary for key columns; key=column_name, value=Found/Not_Found
exclude = {}    # a dictionary for exclude columns; key=column_name, value=Found/Not_Found
hdr = []        # a list for all columns in the header
datadict = {}   # a dictionary for all data lines in one file

# ==============================================================================
# ERROR CODES
# ==============================================================================

e_invalid_args = 1
e_null_key = 2
e_null_exclude = 3
e_key_exclude_overlap = 4
e_key_not_found = 5
e_exclude_not_found = 6
e_headers_mismatch = 7

# The columns of the header line are elements of the array hdr and are objects of
# the following class HeaderCol. A header column has a name, and is either a key or
# an exclude column.
class HeaderCol:
    def __init__(self,name):
        self.name = name
        self.key = False
        self.exclude = False
    def setkey(self):
        self.key = True
    def setexclude(self):
        self.exclude = True

# ==============================================================================
# FUNCTION SECTION
# ==============================================================================

# A function to print the usage of this script
def check_usage():
    if len(sys.argv) == 1:
        print "Usage: python compare_files.py -f file1,file2 -k <KEY> -p <N>"
        print "       <KEY> = comma-delimited list of key columns"
        print "       <N>   = integer degree of precision sought for comparison"
        sys.exit(e_invalid_args)

def process_args():
    global file1, file2, precision
    p = optparse.OptionParser()
    # An option taking an argument
    p.add_option("-f", action="store", dest="file_list")
    p.add_option("-k", action="store", dest="key_list")
    p.add_option("-p", action="store", dest="precision")
    p.add_option("-x", action="store", dest="exclude_list")
    # Parse the command line
    opts, args = p.parse_args()
    # Retrieve the option settings
    (file1, file2) = opts.file_list.split(",")
    for i in opts.key_list.split(","):
        if i == "":
            print "At least one key column is a null string. Abnormal exit."
            sys.exit(e_null_key)
        keycols[i] = "Not_Found"  # assume that key column was not seen in header
    precision = opts.precision
    if precision == None:
        precision = 0.01
    if opts.exclude_list != None:
        for i in opts.exclude_list.split(","):
            if i == "":
                print "At least one exclude column is a null string. Abnormal exit."
                sys.exit(e_null_exclude)
            if i in keycols.keys():
                print "Keys and exclude columns overlap. Abnormal exit."
                sys.exit(e_key_exclude_overlap)
            exclude[i] = "Not_Found" # assume that exclude column was not seen in header

# Read the header line of one file and check that all key columns and exclude columns
# were present in the header
def read_header(file):
    f = open(file)
    line = f.readline()
    f.close()
    for i in line.rstrip("\r\n").split(","):
        a = HeaderCol(i)
        if i in keycols:
            a.setkey()
            keycols[i] = "Found"
        if i in exclude:
            a.setexclude()
            exclude[i] = "Found"
        hdr.append(a)
    # All key columns and exclude columns must be found in the header list
    for item in keycols:
        if keycols[item] == "Not_Found":
            print "Key column: ", item, "was not found in the header. Abnormal exit."
            sys.exit(e_key_not_found)
    for item in exclude:
        if exclude[item] == "Not_Found":
            print "Exclude column: ", item, "was not found in the header. Abnormal exit."
            sys.exit(e_exclude_not_found)

# Read the header of the second file and check that the names and order of the header
# columns match that of the first.
def compare_headers(file):
    f = open(file)
    line = f.readline()
    f.close()
    j = 0
    for i in line.rstrip("\r\n").split(","):
        if j >= len(hdr):
            print "The headers of the two files do not match. Abnormal exit."
            sys.exit(e_headers_mismatch)
        if i != hdr[j].name:
            print "The headers of the two files do not match. Abnormal exit."
            sys.exit(e_headers_mismatch)
        j += 1
    if j < len(hdr):
        print "The headers of the two files do not match. Abnormal exit."
        sys.exit(e_headers_mismatch)

# Read the data of the first file and store it in a hash called datadict. The key is a
# comma-delimited string of the key column values. The value is the line of data read,
# with keys and exclude columns replaced by zero-length strings.
def read_data(file):
    f = open(file)
    first_line = True
    for line in f:
        if first_line == True:
            first_line = False
            continue
        key = ""
        arr = []
        for (j, k) in enumerate(line.rstrip("\r\n").split(",")):
            if j >= len(hdr):
                continue
            if hdr[j].key == True:
                key += "," + k
                arr.append("")
            elif hdr[j].exclude == True:
                arr.append("")
            else:
                arr.append(k)
        key = key.lstrip(",")
        datadict[key] = arr
    f.close

# Iterate through the second file and compare it against the hash datadict.
# If a key is found, compare the non-key columns and print mismatched columns, if found.
# If a key is not found, print the appropriate message.
# Finally, check if a key was found in the first file but not in the second.
def compare_data(iter_file, ref_file):
    # iter_file = file we are iterating through
    # ref_file = file we are referencing against
    f = open(iter_file)
    first_line = True
    for line in f:
        if first_line == True:
            first_line = False
            continue
        key = ""
        arr = []
        for (j, k) in enumerate(line.rstrip("\r\n").split(",")):
            if j >= len(hdr):
                continue
            if hdr[j].key == True:
                key += "," + k
                arr.append("")
            elif hdr[j].exclude == True:
                arr.append("")
            else:
                arr.append(k)
        key = key.lstrip(",")
        if key in datadict:
            # the key was found in both files; compare their non-key data
            mismatches = compare_arrays(arr, datadict[key])
            if mismatches != "":
                print "KEY: %-30s => Mismatched columns = %s" % ("("+key+")", mismatches)
            del datadict[key]
        else:
            print "KEY: %-30s => Found only in %s" % ("("+key+")", iter_file)
    f.close
    for k in datadict:
        print "KEY: %-30s => Found only in %s" % ("("+k+")", ref_file)

# A utility function to compare arrays and return a comma-delimited list of mismatched
# column names and positions.
def compare_arrays(arr1, arr2):
    mm = ""
    for (i, elem) in enumerate(arr1):
        if are_unequal(elem, arr2[i]):
            mm += "," + hdr[i].name + ":" + str(i+1)
    mm = mm.lstrip(",")
    return mm

def are_unequal(value1, value2):
    # Returns True if both values are unequal, False if equal
    # If both values are numbers, then compare them as numbers
    # with values acceptable within degree of precision; otherwise
    # compare them as strings
    if is_num(value1) and is_num(value2):
        diff = abs(float(value1) - float(value2))
        if float(diff) >= float(precision):
            return True
        else:
            return False
    else:
        return value1 != value2

# A function to determine if the parameter is a number or not
def is_num(s):
   try:
       float(s)
       return True
   except ValueError:
       return False

# ==============================================================================
# MAIN SECTION
# ==============================================================================
check_usage()
process_args()
read_header(file1)
compare_headers(file2)
read_data(file1)
compare_data(file2, file1)

File: compare_files.rb
#!/usr/bin/ruby
# ==============================================================================
# Name: compare_files.rb
# Desc: A Ruby program to compare files.
# By:   prat
# On:   11/21/2013
# ==============================================================================
require 'optparse'

E_INVALID_ARGS = 1
E_INVALID_FILENAME = 2
E_INVALID_KEY = 3
E_INVALID_PRECISION = 4
E_INVALID_EXCLUDE = 5
E_KEY_EXCLUDE_OVERLAP = 6
E_INCORRECT_FILE_COUNT = 7
E_KEY_NOT_FOUND = 8
E_EXCLUDE_NOT_FOUND = 9
E_UNEQUAL_HEADERS = 10
$dp = 0.01
$keys = Hash.new
$exclude = Hash.new
$hdr = Array.new
$data = Hash.new

# ==============================================================================
# CLASS DEFINITIONS
# ==============================================================================
class Header
  def initialize(name)
    @name = name
    @iskey = false
    @isexclude = false
  end
  def name; @name; end
  def iskey; @iskey; end
  def isexclude; @isexclude; end
  def iskey=(value)
    @iskey = value
  end
  def isexclude=(value)
    @isexclude = value
  end
  def to_s                           # Return a String that represents this header
    "(#@name, #@iskey, #@isexclude)" # Just interpolate the instance variables into a string
  end
end

# ==============================================================================
# GLOBAL FUNCTIONS
# ==============================================================================
def is_number?(object)
  true if Float(object) rescue false
end

def process_args
  OptionParser.new do |o|
    o.on("-f", "--files FILE_LIST", "Comma-delimited list of files to compare") do |files|
      $files = files.split(",")
      $files.each do |x|
        if x =~ /^-/ or x == ""
          print "Invalid file name!\n"
          exit E_INVALID_FILENAME
        end
      end
    end
    o.on("-k", "--keys KEY_LIST", "Comma-delimited list of keys") do |keylist|
      keylist.split(",").each do |k|
        if k =~ /^-/ or k == ""
          print "Invalid key column!\n"
          exit E_INVALID_KEY
        end
        $keys[k] = 0
      end
    end
    o.on("-p", "--precision N", "Degree of precision") do |n|
      $dp = n
      if not is_number?($dp)
        print "Invalid degree of precision!\n"
        exit E_INVALID_PRECISION
      end
    end
    o.on("-x", "--exclude EXCLUDE_LIST", "Comma-delimited list of columns to be excluded from comparison") do |xlist|
      xlist.split(",").each do |x|
        if x =~ /^-/ or x == ""
          print "Invalid exclude column!\n"
          exit E_INVALID_EXCLUDE
        end
        $exclude[x] = 0
      end
    end
    o.on("-h", "--help", "Show this message") do
      puts o
      exit E_INVALID_ARGS
    end
    if ARGV.length == 0
      puts o
      exit E_INVALID_ARGS
    end
    o.parse!
  end
  # Balk if key and exclude columns overlap
  $keys.each_key do |x|
    if $exclude.has_key?(x)
      print "Keys and exclude columns overlap!\n"
      exit E_KEY_EXCLUDE_OVERLAP
    end
  end
  # Or if number of files is not 2
  if $files.size != 2
    print "Incorrect number of files specified!\n"
    exit E_INCORRECT_FILE_COUNT
  end
end

def read_header(file)
  File.open(file) do |f|
    f.readline.chomp!.split(",").each do |x|
      a = Header.new(x)
      if $keys.has_key?(x)
        a.iskey = true
        $keys[x] = 1
      end
      if $exclude.has_key?(x)
        a.isexclude = true
        $exclude[x] = 1
      end
      # puts a
      $hdr << a
    end
  end
  # All keys and exclude cols must be present in the header
  $keys.each do |k, v|
    if v == 0
      print "Key ", k, " was not found in the header!\n"
      exit E_KEY_NOT_FOUND
    end
  end
  $exclude.each do |k, v|
    if v == 0
      print "Exclude column ", k, " was not found in the header!\n"
      exit E_EXCLUDE_NOT_FOUND
    end
  end
end

def compare_headers(file)
  arr1 = Array.new
  arr2 = Array.new
  $hdr.each {|x| arr1 << x.name}
  File.open(file) {|f| arr2 = f.readline.chomp!.split(",")}
  if (arr1 <=> arr2) != 0
    print "The headers of the two files do not match!\n"
    exit E_UNEQUAL_HEADERS
  end
end

def read_data(file)
  arr = Array.new
  key = ""
  fh = File.open(file)
  fh.each do |line|
    next if $. == 1
    line.chomp!
    line.split(",").each_with_index do |token, pos|
      next if pos >= $hdr.size
      if $hdr[pos].iskey
        key << "," + token
        arr << ""
      elsif $hdr[pos].isexclude
        arr << ""
      else
        arr << token
      end
    end
    key[0] = ""
    $data[key] = arr
    key, arr = "", []
  end
end

def compare_data(files)
  arr = Array.new
  key = ""
  iter, ref = files    # iter = file to iterate through; ref = reference file
  fh = File.open(iter)
  fh.each do |line|
    next if $. == 1
    line.chomp!
    line.split(",").each_with_index do |token, pos|
      next if pos >= $hdr.size
      if $hdr[pos].iskey
        key << "," + token
        arr << ""
      elsif $hdr[pos].isexclude
        arr << ""
      else
        arr << token
      end
    end
    key[0] = ""
    # if key exists in data hash then compare non-key values,
    # and print message only if any non-key columns are different
    # if key does not exist in data hash then print message
    if $data.has_key?(key)
      mismatches = compare_arrays(arr, $data[key])
      if mismatches != ""
        printf("KEY: %-30s => Mismatched columns = %s\n", "(" + key + ")", mismatches)
      end
      $data.delete(key)
    else
      printf("KEY: %-30s => Found only in %s\n", "(" + key + ")", iter)
    end
    key, arr = "", []
  end
  # the remaining keys in data hash were found only in the ref file
  $data.keys.each do |k|
    printf("KEY: %-30s => Found only in %s\n", "(" + k + ")", ref)
  end
end

def compare_arrays(arr1, arr2)
  str = ""
  arr1.each_with_index do |elem, i|
    if is_number?(elem) and is_number?(arr2[i])
      str << "," + $hdr[i].name + ":" + (i+1).to_s if (Float(elem) - Float(arr2[i])).abs >= Float($dp)
    else
      str << "," + $hdr[i].name + ":" + (i+1).to_s if elem != arr2[i]
    end
  end
  str[0] = ""
  return str
end

# ==============================================================================
# MAIN SECTION
# ==============================================================================
process_args
read_header($files[0])
compare_headers($files[1])
read_data($files[0])
compare_data($files.values_at(1,0))

